nodejs/eks/authenticationMode.ts

```diff
@@ -0,0 +1,192 @@
+
+export const CONFIG_MAP = "CONFIG_MAP";
+export const API_AND_CONFIG_MAP = "API_AND_CONFIG_MAP";
+export const API = "API";
+
+export function validateAuthenticationMode(args: ClusterOptions) {
+    if (
+        args.authenticationMode &&
+        args.authenticationMode !== CONFIG_MAP &&
+        args.authenticationMode !== API_AND_CONFIG_MAP &&
+        args.authenticationMode !== API
+    ) {
+        throw new Error(
+            `Invalid value for authenticationMode: ${args.authenticationMode}. Allowed values are: ${CONFIG_MAP}, ${API_AND_CONFIG_MAP}, ${API}.`,
+        );
+    }
+
+    const configMapOnlyProperties: (keyof ClusterOptions)[] = [
+        "roleMappings",
+        "userMappings",
+        "instanceRoles",
+        "instanceRole",
+    ];
+    const apiOnlyProperties: (keyof ClusterOptions)[] = ["accessEntries"];
+
+    if (!supportsConfigMap(args.authenticationMode)) {
+        configMapOnlyProperties.forEach((prop) => {
+            if (args[prop]) {
+                throw new Error(
+                    `The '${prop}' property is not supported when 'authenticationMode' is set to '${args.authenticationMode}'.`,
+                );
+            }
+        });
+    }
+
+    if (!supportsAccessEntries(args.authenticationMode)) {
+        apiOnlyProperties.forEach((prop) => {
+            if (args[prop]) {
+                const errorMsg = args.authenticationMode != null ? `set to '${args.authenticationMode}` : "not set";
+                throw new Error(
+                    `The '${prop}' property is not supported when 'authenticationMode' is ${errorMsg}.`,
+                );
+            }
+        });
+    }
+}
+
+export function supportsConfigMap(authenticationMode: string | undefined) {
+    // If authenticationMode is not provided, it defaults to CONFIG_MAP
+    return (
+        !authenticationMode ||
+        authenticationMode === CONFIG_MAP ||
+        authenticationMode === API_AND_CONFIG_MAP
+    );
+}
+
+export function supportsAccessEntries(authenticationMode: string | undefined) {
+    return authenticationMode === API || authenticationMode === API_AND_CONFIG_MAP;
+}
+
+/**
+ * Creates the AWS authentication data for the aws-auth ConfigMap.
+ *
+ * @param instanceRoles - The instance roles to be mapped.
+ * @param roleMappings - The IAM role mappings to be included.
+ * @param userMappings - The IAM user mappings to be included.
+ * @returns The AWS authentication data for the aws-auth ConfigMap.
+ * @throws Error if the IAM role mappings or user mappings are invalid or cannot be serialized to YAML.
+ */
+export function createAwsAuthData(
+    instanceRoles: pulumi.Output<aws.iam.Role[]>,
+    roleMappings: pulumi.Input<pulumi.Input<RoleMapping>[]> | undefined,
+    userMappings: pulumi.Input<pulumi.Input<UserMapping>[]> | undefined,
+): pulumi.Input<{ [key: string]: pulumi.Input<string> }> {
+    const instanceRoleMappings = instanceRoles.apply((roles) =>
```

Comment:
This is the same as before, just extracted it to here to slim down the `cluster.ts` a bit and make it more manageable